### PR TITLE
abscomp_colm

### PR DESCRIPTION
**Pull request by profxj**
A set of methods related to column densities:
  Apparent column density plot
  Synthesize multiple column density measurements
  COG

And related tests and documentation.

### CODE REVIEW

File: docs/abscomp.rst [hunk 1]
```diff
@@ -70,3 +72,41 @@ a stack plot (aka velocity plot) is generated with::
...
+are combined in a weighted mean whereas limits are compared
+and the most sensitive one is adopted.::
+
+    abscomp.synthesize_colm()
+
+Here is the set of rules:
+
+1.  If all measurements are upper limits, take the lowest value and flag as an upper limit (*flgN=3*).
+2.  If all measurements are a mix of upper and lower limits, take the highest lower limit and flag as a lower limit (*flgN=2*).
+3.  If one or measurements are a proper detection, take the weighted mean of these and flag as a detection (*flgN=1*).
```
Comment (nhmc) [inline docs/abscomp.rst]: typo 'if one or more'

File: linetools/analysis/absline.py [hunk 2]
```diff
@@ -5,29 +5,33 @@
...
+
+# Atomic constant
 atom_cst = (const.m_e.cgs*const.c.cgs / (np.pi * 
     (const.e.esu**2).cgs)).to(u.AA*u.s/(u.km*u.cm**2))
 
 # Perform AODM on the line
 def aodm(spec,idata):
-    """  AODM calculation 
+    """  AODM calculation on an absorption line
     See Savage & Sembach 1991, ApJ, 379, 245
```
Comment (nhmc) [inline linetools/analysis/absline.py]: Please a leave a blank line between the single line summary (here """ AODM calculation on an absorption line) and any extra text below.  This makes Sphinx happy.

File: linetools/analysis/absline.py [hunk 3]
```diff
@@ -69,3 +73,74 @@ def aodm(spec,idata):
 
     # Return
     return ntot, np.sqrt(tvar), flg_sat
+
+
+def log_clm(obj):
+    """Return logN and sig_logN given linear N, sigN
```
Comment (nhmc) [inline linetools/analysis/absline.py]: Need a blank line here.

File: linetools/analysis/absline.py [hunk 4]
```diff
@@ -69,3 +73,74 @@ def aodm(spec,idata):
...
+    Parameters
+    ---------
```
Comment (nhmc) [inline linetools/analysis/absline.py]: Also dashes must be the same length as the heading for Sphinx (Sorry - this is a bit annoying, I know)

File: linetools/analysis/absline.py [hunk 5]
```diff
@@ -69,3 +73,74 @@ def aodm(spec,idata):
...
+    sig_logN = np.sqrt(lgvar)
+    # Return
+    return logN, sig_logN
+
+def sum_logN(obj1,obj2):
+    """Add log columns and return value and errors
+    Parameters
+    ----------
+    obj 1: object
+      An object with tags appropriate for the analysis
```
Comment (nhmc) [inline linetools/analysis/absline.py]: I think we should use 'keys' here instead of 'tags'

File: linetools/analysis/cog.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+# Now interpolate
+intFtau0 = interp1d(all_tau0, xFtau0, bounds_error=False,fill_value=0.)
+
+##############################
+def cog_plot(COG_dict):
+    """Generate a plot for COG solution
+    Parameters
+    ----------
+    COG_dict : dict
+      dict containing the COG inputs and solution
```
Comment (nhmc) [inline linetools/analysis/cog.py]: reference function that provides the COG solution

File: linetools/isgm/abscomponent.py [hunk 7]
```diff
@@ -167,6 +174,167 @@ def build_table(self):
...
+        # Legend
+        legend = ax.legend(loc='upper left', scatterpoints=1, borderpad=0.3,
+                               handletextpad=0.3, fontsize='large')
+
+        plt.tight_layout(pad=0.2,h_pad=0.,w_pad=0.1)
+        if show:
+            plt.show()
+        plt.close()
+
+    def synthesize_colm(self, clobber=False, redo_aodm=False, **kwargs):
```
Comment (nhmc) [inline linetools/isgm/abscomponent.py]: I prefer 'overwrite' instead of 'clobber'

File: linetools/isgm/abscomponent.py [hunk 8]
```diff
@@ -167,6 +174,167 @@ def build_table(self):
...
+
+    def plot_Na(self, show=True, **kwargs):
+        """Plot apparent column density Na profiles
+        """
+        import matplotlib.pyplot as plt
+        import matplotlib.gridspec as gridspec
+        import matplotlib as mpl
+        try: # Nicer view, especially in notebook
+            import seaborn as sns; sns.set(context="notebook",font_scale=2)
+        except:
```
Comment (nhmc) [inline linetools/isgm/abscomponent.py]: maybe change to `except ImportError:`. Catching all exceptions is seen as bad style among Python people.

File: linetools/analysis/cog.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    # Data with values
+    gdv = COG_dict['redEW'] > 0.
+    yerr=(COG_dict['sigEW'][gdv]/COG_dict['wrest'][gdv])/COG_dict['redEW'][gdv]
+    ax.errorbar(np.log10(COG_dict['f'][gdv]*COG_dict['wrest'][gdv].to('cm').value),
+                np.log10(COG_dict['redEW'][gdv]),
+                yerr=yerr, fmt='o')
+    # Upper limit
+    upper = COG_dict['redEW'] <= 0.
+    ax.scatter(np.log10(COG_dict['f'][upper]*COG_dict['wrest'][upper].to('cm').value),
+               np.log10(3*COG_dict['sigEW'][upper]/COG_dict['wrest'][upper]), color='red', marker='v')
```
Comment (nhmc) [inline linetools/analysis/cog.py]: This is fine for upper limits. But you can also add them using plt.errorbar and they look nicer: http://matplotlib.org/examples/statistics/errorbar_limits.html

**Comment by nhmc** [comment]
This is great, thank you!

I will have to try it out before I have a thorough idea of how everything works, but I'm happy to merge this now.

**Comment by nhmc** [comment]
(please merge whenever you're ready)

File: docs/abscomp.rst (hunk 1 above)
Comment (profxj) [inline docs/abscomp.rst]: Fixed

File: linetools/analysis/absline.py (hunk 2 above)
Comment (profxj) [inline linetools/analysis/absline.py]: Done

File: linetools/analysis/absline.py (hunk 3 above)
Comment (profxj) [inline linetools/analysis/absline.py]: Added

File: linetools/analysis/absline.py (hunk 4 above)
Comment (profxj) [inline linetools/analysis/absline.py]: Yup, I knew that one.  My bad.

File: linetools/analysis/absline.py (hunk 5 above)
Comment (profxj) [inline linetools/analysis/absline.py]: Agreed

File: linetools/analysis/cog.py (hunk 6 above)
Comment (profxj) [inline linetools/analysis/cog.py]: ok

File: linetools/isgm/abscomponent.py (hunk 7 above)
Comment (profxj) [inline linetools/isgm/abscomponent.py]: done

File: linetools/isgm/abscomponent.py (hunk 8 above)
Comment (profxj) [inline linetools/isgm/abscomponent.py]: Agreed.  Was too lazy to figure out which one I needed.

File: linetools/analysis/cog.py (hunk 9 above)
Comment (profxj) [inline linetools/analysis/cog.py]: While I agree, I don't have the endurance for this one..

**Comment by profxj** [comment]
Merging.  Will add in error analysis for COG once
I understand it.

**Pull request merged**